kernel/signal.c

```diff
@@ -756,9 +756,17 @@ long myst_signal_clone(
     return ret;
 }
 
-long myst_handle_host_signal(siginfo_t* siginfo, mcontext_t* mcontext)
+void myst_handle_host_signal(siginfo_t* siginfo, mcontext_t* mcontext)
 {
-    return _handle_one_signal(siginfo->si_signo, siginfo, mcontext);
+    _handle_one_signal(siginfo->si_signo, siginfo, mcontext);
+
+    /* The signal handler in the programming language runtime or the application
+     * might not return. If execution returns here, set the CPU context as the
+     * returned mcontext to continue execution. */
+    myst_sigreturn(mcontext);
+
+    // Unreachable
+    assert(0);
```

Comment:
The exception handler is invoked by this line in OE:

https://github.com/openenclave/openenclave/blob/e492526a243c7836a7993c48cb6fb3d0b7c5509d/enclave/core/sgx/exception.c#L296

OE expects that control returns from the handler, and performs cleanup after the return from exception handlers:
https://github.com/openenclave/openenclave/blob/e492526a243c7836a7993c48cb6fb3d0b7c5509d/enclave/core/sgx/exception.c#L304-L310

If the exception handler does not return, then the cleanup won't be executed and would leave OE in an inconsistent state.



T

Reply:
That was the behavior before this PR:
OE exception handler -> Mystikos enclave exception handler -> Mystikos kernel signal handler -> Mystikos user signal handler -> (return) Mystikos kernel signal handler -> (return) OE exception handler -> resume execution

With this PR, the flow is updated to:
OE exception handler -> Mystikos enclave exception handler -> (return) OE exception handler -> Mystikos kernel signal handler -> Mystikos user signal handler -> (optional return) Mystikos kernel signal handler -> resume execution

In summary, at this point, the OE second-pass handler has returned and done the cleanup already.